synapse/app/homeserver.py

```diff
@@ -190,7 +191,8 @@ def _configure_named_resource(self, name, compress=False):
                     "/_matrix/client/unstable": client_resource,
                     "/_matrix/client/v2_alpha": client_resource,
                     "/_matrix/client/versions": client_resource,
-                    "/.well-known/matrix/client": WellKnownResource(self),
+                    "/.well-known/matrix/client": well_known_resource,
+                    "/.well-known/matrix/server": well_known_resource,
```

Comment:
I think it would be better to use different `Resource` classes than a single one which then has to do path-matching to figure out which type of well-known to serve.